### ISSUE TITLE
[Feature Request] Update docker-compose stacks and/or Respect docker-compose "depends_on"

### ISSUE THREAD

**Issue by carlyman** [mag37/dockcheck#49]
When updating a set of containers within a docker-compose file, it should respect depends_on within a given service. Restarting each container individually can lead to overall issues in the stack working properly.  Alternatively, choosing to stop a docker-compose stack, updating, and restarting would work as well.

**Comment by mag37** [comment]
I'm not entirely sure I get what you mean.

It does update the images individually, even though they're in a stack with others. This is because sometimes people want to selectively update specific images - if it would just `docker compose pull` it would update all images, including those specifically skipped while picking.

This could probably be built differently with more logic around it to handle it smoother, but I'll have to do some thinking if so. 

But to your point about not respecting `depends_on`, if you mean it doesn't do that at all? - I'll have to do some testing. Because it's just `docker compose up -d` so it should still start appropriately in the right order.

**Comment by carlyman** [comment]
If a stack has 3 containers, one gets updated, and just that single container gets restarted...I've had issues with functionality at times. (of course, this could be poor docker container designs).  When manually updating, I need to ensure all containers within that stack are stopped first before restarting to ensure they start in proper order.

**Comment by mag37** [comment]
You're correct, I just did some testing on how docker handle the restart when you only update one in a stack. But I think this is deliberate. 

I had a compose file with ContainerA depending on ContainerB.

I pulled a new image to ContainerA, did `docker compose up -d` and it only recreated ContainerA while B were kept running.

While with the same setup (A depending on B), I pulled a new image to B and `docker compose up -d` and it recreated both; First B then A. 

Hence I think it do respect the `depends_on` logic?

**Comment by carlyman** [comment]
`depends_on` is probably the wrong model....in my use-case, i need to `docker compose stop` rather than just `docker compose up -d` after pulling a new image. Would it be possible to stop the docker stack and then restarting them all after any image within that stack is updated?

**Comment by mag37** [comment]
I was considering that as the best solution, just pulling an image, then instead of just `up -d`, first stop and then up.

That would mean unnecessary stops but maybe it will solve more issues that way.

**Comment by mag37** [comment]
So that would be changing from:

```bash
      if [ -n "$ContEnv" ]; then 
        $DockerBin ${CompleteConfs[@]} --env-file "$ContEnv" up -d "$ContName"
      else
        $DockerBin ${CompleteConfs[@]} up -d "$ContName"
      fi
```

To:
```bash
      if [ -n "$ContEnv" ]; then 
        $DockerBin stop && $DockerBin ${CompleteConfs[@]} --env-file "$ContEnv" up -d "$ContName"
      else
        $DockerBin stop && $DockerBin ${CompleteConfs[@]} up -d "$ContName"
      fi
```

I'll do some testing and see if there's any implications. 

Merry Christmas!🎄

**Comment by mag37** [comment]
I've been considering the pros and cons of implementing this.. And while I've never encountered issues with this directly I can see the point. 
The downside though is that every compose-stack will be restarted X amount of times depending on X amount of images within it.
Eg. if you run `dockcheck-sh -y` to auto update your whole fleet and a compose file contains 5 different containers/images - that compose stack would be restarted for every container/image within it - 5 times.

To avoid this a whole new level of tracking what's been updated and not is required.
I need to think further on this one.

**Comment by mag37** [comment]
Read this in a different thread, worth considering.
>There's been a move away from specifying container dependencies in compose. They're only valid at startup time and don't work when dependent containers are restarted at run time. Instead, each container should include mechanism to retry to reconnect to dependent services when the connection is dropped. Many libraries to connect to databases or REST API services have configurable built-in retries. I'd look into that. It is needed for production code anyway.

[Source](https://stackoverflow.com/a/47714157)

**Comment by carlyman** [comment]
For sure appreciate not a single ideal approach and that there may be better docker stack hygiene to ensure containers work without stopping the whole stack.  

Perhaps having an Environmental variable to "stop stack first" before upgrading is a compromise as some stacks out there do require it?

Thanks for the consideration (and Merry Christmas to you, too!)

**Comment by mag37** [comment]
I'm not sure how to continue with this - I might leave it for now but maybe in the future add another option flag like `-f` **force stop+start a stack if image is updated** or something.
But until then if you want to modify your own script to restart on every updated image - just change the lines I wrote above (line 283 and 285 in the actual code atm).

Or let me know if you've got a different suggestion to solve this :)
